### PR TITLE
[FIX #450] Remove creating global catalog variable from status-go.

### LINKED ISSUE #450: Remove creating global catalog variable from status-go
# Problem

During a jail cell creation and initialization, a JS code is executed that looks out of place:
```go
func (j *Jail) CreateAndInitCell(chatID string, code ...string) string {
  // create and init a cell...

  // this is inlined
  _, err := cell.Run(`var catalog = JSON.stringify(_status_catalog)`)
  if err != nil {
    return newJailErrorResponse(err)
  }

  // return `catalog` as string
}
```

As `_status_catalog` is not defined in status-go, this code should not be here as well.

# Implementation

~~When PR #401 is merged, there will be a new binding `ExecuteJS` which will allow running this JS code from status-react.~~

~~Prerequisite: contact status-react team before starting working on this and iron out the final details.~~

I talked to status-react team and we don't want to create `ExecuteJS` binding just for this detail. This binding should be removed. Instead, we want to return the last result of `cell.Run(js)` executed inside `Jail.createAndInitCell()` method.

# Acceptance Criteria

1. status-react team is aware of this change and committed to fixing it on their side,
1. JS code creating `catalog` variable is eliminated from status-go,
1. Tests that check `CreateAndInitCell` are fixed.

### PR DESCRIPTION
**Pull request by mandrigin**
I tried to make on both sides: status-react and status-go.
This change is trivial (thanks to @divan for pre-refactoring).

I'll keep this one as `WIP` until [status-react#2587](https://github.com/status-im/status-react/pull/2587) is merged.

Closes #450

### CODE REVIEW

**Comment by mandrigin** [comment]
Fixed `./e2e/jail/` tests, some more to go.

**Comment by mandrigin** [comment]
Fixed `test-e2e`.

File: geth/jail/jail.go
Comment (adambabik) [inline geth/jail/jail.go]: By default, Go returns compact JSON, so this should be `{"result":"success"}`.

**Review by adambabik — APPROVED** [review]
Great job! 

Let's wait with merging this PR until someone from status-react confirms it's ok on their side.

File: geth/jail/jail.go
Comment (mandrigin) [reply geth/jail/jail.go]: updated json to the compact notation

**Comment by mandrigin** [comment]
Thanks for reviewing! Yes, let’s wait for the react part to be discussed/merged first.

**Comment by adambabik** [comment]
@mandrigin sorry about a slight change of plans but we need to implement it a bit differently. Take a look at updated "Implementation" description in https://github.com/status-im/status-go/issues/450 and let me know if you have any questions. 

Basically, `CreateAndInitCell` takes variadic `code` arguments. We want to return the last result of executed JS code, so the last result of `_, err := cell.Run(js)`, which now is ignored. The result should be returned as a string. The usage on status-react side will look like this:
```
CreateAndInitCell('some-cell-id', `var catalog = JSON.stringify(_status_catalog); catalog`);
```
or
```
CreateAndInitCell('some-cell-id', `var catalog = JSON.stringify(_status_catalog)`, `catalog`);
```

If there is no `code` arguments, let's return an empty string.

**Comment by mandrigin** [comment]
@adambabik sure that makes total sense! I'll update it.

I have one question, is there any special meaning on why there is a separate variable: `catalog`?
As I see things
```go
CreateAndInitCell('some-cell-id', `var catalog = JSON.stringify(_status_catalog)`, `catalog`);
```
and
```go
CreateAndInitCell('some-cell-id', `JSON.stringify(_status_catalog)`);
```
are absolutely equal.

**Comment by adambabik** [comment]
@mandrigin thanks! 

With `var catalog = JSON.stringify(_status_catalog)` a global `catalog` variable is created in the JS runtime environment. I believe status-react uses that variable later or Dapps have access to it.

**Comment by mandrigin** [comment]
@adambabik updated `CreateAndInitCell` and `Parse` behaviours.

File: geth/jail/jail.go
Comment (mandrigin) [inline geth/jail/jail.go]: I'm a bit hesitant about returning 3 values here, but it seems like the simplest solution.
We use both `Cell` and `result` in the callers'.

**Comment by adambabik** [comment]
Thanks @mandrigin! LGTM

**Comment by mandrigin** [comment]
I opened a new PR in status-react to support this change: https://github.com/status-im/status-react/pull/2671

After it is merged, I will resolve `WIP` status of this one.

**Comment by adambabik** [comment]
@mandrigin As https://github.com/status-im/status-react/issues/2530 has been resolved, can you please resolve conflicts here so that we can merge this PR as well?

**Comment by mandrigin** [comment]
Sure! I'll try to do that today!

**Comment by mandrigin** [comment]
@adambabik it took some time, but it seems to be done now :)

File: e2e/jail/jail_test.go
Comment (adambabik) [inline e2e/jail/jail_test.go]: FYI `;` are not needed at the end of JS code. It should properly concatenate anyway.

File: e2e/jail/jail_test.go
Comment (adambabik) [inline e2e/jail/jail_test.go]: This binding should not be used because it will be removed. Let's call like this:
```go
response = s.Jail.CreateAndInitCell(testChatID, `var test2 = 'bar'`, `test2`)
```
 
After that, we can deduplicate some calls below.

File: e2e/jail/jail_test.go
Comment (adambabik) [inline e2e/jail/jail_test.go]: This test name is wrong now as it does not test any errors, right? Let's merge this, `TestCreateAndInitCellResponse` and `TestCreateAndInitCellResponseWithJS` into a single test case called `TestCreateAndInitCell`.

File: lib/utils.go
Comment (adambabik) [inline lib/utils.go]: I would suggest swapping the order of `jailFunc` so that `Name` goes first. It will be easier to reason about otherwise it suggests some params to function passed in the first field.

File: lib/utils.go
Comment (adambabik) [inline lib/utils.go]: I am not sure if this helper is needed. The biggest assumption of it is that it all jail functions have the same signature which may be true now but won't have to be tomorrow.

Maybe we could just replace `reflect.DeepEqual` call with `require.Equal`. It could look like this:

```go
response := CreateAndInitCell(
  C.CString("CHAT_ID_INIT_TEST"), 
  C.CString(`var extraFunc = function (x) { return x * x; }; extraFunc(2);`),
)
require.Equal(t, `{"result": 4}`, C.GoString(response))
```

I think it's not too much code and it's pretty easy to understand.

**Review by adambabik — CHANGES_REQUESTED** [review]

File: lib/utils.go
Comment (mandrigin) [reply lib/utils.go]: I agree with you, the whole idea of this refactoring was to make Code Climate happy.
It currently has checks for max 25 LoC per function and for max 4 return statements.

I think these limits make sense for the code itself, but maybe they shouldn't be that strict with test methods.

Comment (adambabik) [reply lib/utils.go]: I am not sure we should care about Code Climate now. This whole file needs heavy refactoring and using `require.Equal` should make it happier in terms of different code paths anyway.

Comment (mandrigin) [reply lib/utils.go]: Ah, 👌 
I thought it would block the merging process.

File: e2e/jail/jail_test.go
Comment (mandrigin) [reply e2e/jail/jail_test.go]: I  wanted to make sure the JS state is fully preserved when the cell is re-initialized. 
Is there a better, non-deprecated way to do that?
Like to initialize another variable using any kind of js call functionality.

**Comment by mandrigin** [comment]
@adambabik fixed the code review notes

**Review by adambabik — APPROVED** [review]

**Comment by adambabik** [comment]
Thanks @mandrigin, LGTM!

File: geth/jail/jail.go
Comment (divan) [inline geth/jail/jail.go]: This function looks like a good candidate to be an internal method of `Cell` type, rather than `Jail`. And to stress the difference with `(*Cell).Run()` maybe renamed to something more explicit for this case (runBulkJS? runInitJS? etc).

File: geth/jail/jail.go
Comment (divan) [inline geth/jail/jail.go]: So the result will be only from the last JS code string?

Comment (mandrigin) [reply geth/jail/jail.go]: Yes. And the Closure counterpart is already tweaked to support this behaviour: [status-react#2587](https://github.com/status-im/status-react/pull/2587)

File: geth/jail/jail.go
Comment (divan) [inline geth/jail/jail.go]: Sorry if it was discussed before, but maybe just `EmptyResponse`?
Creating cell is currently the only place where this response is returned, so no need to duplicate "cell created" in error name. Moreover, theoretically this empty response can be used for other actions, so storing "cell created" in it will make even less sense.

**Review by divan — COMMENTED** [review]
A couple of remarks.

File: geth/jail/jail.go
Comment (mandrigin) [reply geth/jail/jail.go]: Actually, I figured out that I can use `cell.Run` for that (and that is used in the method just above that).

**Comment by mandrigin** [comment]
@divan fixed the remarks, code climate and Travis are happy too 🎉

**Review by divan — APPROVED** [review]
LGTM

**Pull request merged**
